var/spack/repos/builtin/packages/netcdf-cxx4/package.py

```diff
@@ -23,7 +26,13 @@ class NetcdfCxx4(AutotoolsPackage):
     force_autoreconf = True
 
     def configure_args(self):
-        return ['CPPFLAGS=-I' + self.spec['netcdf'].prefix.include]
+        cflags = fflags = ''
+        if '+pic' in self.spec:
```

Comment:
Hi @scheibelp, thanks for the reply.
I'm not yet so familiar with spack to understand all the details of the documentation you linked; but I got that setting `CFLAGS` in `configure_args` as I did in the netcdf* packages overrides compiler flags.

Using a `setup_environment` method as in the zlib package could fix this?
https://github.com/spack/spack/blob/eea786f4e888b30cf70d1cddab4b5e6aa1c4cc28/var/spack/repos/builtin/packages/zlib/package.py#L41-L46

Reply:
If you use `spack_env.append_flags` in `setup_environment` it will not conflict with compiler flags set in `compilers.yaml`. It may be important to set these as arguments to the build system though (as was done for `netcdf` originally.

Reply:
So moving from the `configure_args` method to a `setup_environment` should fix the problem, right?
Something like:
```python
 def setup_environment(self, spack_env, run_env): 
     if '+pic' in self.spec: 
         spack_env.append_flags('CFLAGS', self.compiler.pic_flag)
         spack_env.append_flags('FFLAGS', self.compiler.pic_flag)

     spack_env.append_flags('CPPFLAGS', '-I' + self.spec['netcdf'].prefix.include) 
```

I don't get what you mean with:
> It may be important to set these as arguments to the build system though (as was done for netcdf originally).

sorry.

Reply:
> I don't get what you mean with...

There are a few ways to set compiler flags for a build

* The spack compiler wrappers can inject them
* They can be expressed as arguments to the build system (configure or CMake)
* They can be exported as environment variables

Build systems may be able to recognize the third case, it depends on how the configure script is written. The `netcdf` package in Spack currently follows the second option.

Moving flag settings from `configure_args` to `setup_environment` will avoid conflicts with `compilers.yaml` but may conflict with the configure script (e.g. if `configure` decides on some optimization flag which is set in `compilers.yaml`)

FWIW you don't have to worry about it for `netcdf` (you can set all the flags as `configure_args`). I brought it up because you asked about `flag_handler` elsewhere.

Reply:
Okay, now I see your point.
So, what's the best way to set compiler flags, in order to avoid conflicts with variables defined in `compilers.yaml`? I could fix the netcdf packages.

Reply:
If you define a `flag_handler` method for each `netcdf` package, you can tell it to export the flags as arguments to the `configure` script. It would look something like:

```
def flag_handler(self, name, flags):
    if name == 'cppflags':
        flags.append('-I' + self.spec['netcdf'].prefix.include)
    elif name in ['fflags', 'cflags']:
        flags.append(self.compiler.pic_flag)
    return (None, None, flags)
```

The `Autotools` package defines a `flags_to_build_system_args` function which will express compiler flags as arguments to the build system, specifically those returned as part of the 3rd element of the `flag_handler` tuple (as mentioned in https://spack.readthedocs.io/en/latest/packaging_guide.html#compiler-flags).